File: java-commons/concurrent/src/main/java/cn/wjybxx/concurrent/UncancellableToken.java
/*
 * Copyright 2023-2024 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.wjybxx.concurrent;

import java.util.concurrent.Executor;
import java.util.function.BiConsumer;
import java.util.function.Consumer;

/**
 * @author wjybxx
 * date - 2024/1/9
 */
final class UncancellableToken implements ICancelToken {

    static final UncancellableToken INST = new UncancellableToken();

    private UncancellableToken() {
    }

    @Override
    public ICancelToken asReadonly() {
        return this;
    }

    // region code

    @Override
    public int cancelCode() {
        return 0;
    }

    @Override
    public boolean isCancelling() {
        return false;
    }

    @Override
    public int reason() {
        return 0;
    }

    @Override
    public int degree() {
        return 0;
    }

    @Override
    public boolean isInterruptible() {
        return false;
    }

    @Override
    public boolean isWithoutRemove() {
        return false;
    }

    @Override
    public void checkCancel() {

    }

    // endregion

    // region 监听器

    private static final IRegistration TOMBSTONE = () -> {};

    @Override
    public IRegistration thenAccept(Consumer<? super ICancelToken> action, int options) {
        return TOMBSTONE;
    }

    @Override
    public IRegistration thenAccept(Consumer<? super ICancelToken> action) {
        return TOMBSTONE;
    }

    @Override
    public IRegistration thenAcceptAsync(Executor executor, Consumer<? super ICancelToken> action) {
        return TOMBSTONE;
    }

    @Override
    public IRegistration thenAcceptAsync(Executor executor, Consumer<? super ICancelToken> action, int options) {
        return TOMBSTONE;
    }

    @Override
    public IRegistration thenAccept(BiConsumer<? super IContext, ? super ICancelToken> action, IContext ctx, int options) {
        return TOMBSTONE;
    }

    @Override
    public IRegistration thenAccept(BiConsumer<? super IContext, ? super ICancelToken> action, IContext ctx) {
        return TOMBSTONE;
    }

    @Override
    public IRegistration thenAcceptAsync(Executor executor, BiConsumer<? super IContext, ? super ICancelToken> action, IContext ctx) {
        return TOMBSTONE;
    }

    @Override
    public IRegistration thenAcceptAsync(Executor executor, BiConsumer<? super IContext, ? super ICancelToken> action, IContext ctx, int options) {
        return TOMBSTONE;
    }

    @Override
    public IRegistration thenRun(Runnable action, int options) {
        return TOMBSTONE;
    }

    @Override
    public IRegistration thenRun(Runnable action) {
        return TOMBSTONE;
    }

    @Override
    public IRegistration thenRunAsync(Executor executor, Runnable action) {
        return TOMBSTONE;
    }

    @Override
    public IRegistration thenRunAsync(Executor executor, Runnable action, int options) {
        return TOMBSTONE;
    }

    @Override
    public IRegistration thenRun(Consumer<? super IContext> action, IContext ctx, int options) {
        return TOMBSTONE;
    }

    @Override
    public IRegistration thenRun(Consumer<? super IContext> action, IContext ctx) {
        return TOMBSTONE;
    }

    @Override
    public IRegistration thenRunAsync(Executor executor, Consumer<? super IContext> action, IContext ctx) {
        return TOMBSTONE;
    }

    @Override
    public IRegistration thenRunAsync(Executor executor, Consumer<? super IContext> action, IContext ctx, int options) {
        return TOMBSTONE;
    }

    @Override
    public IRegistration thenNotify(CancelTokenListener action, int options) {
        return TOMBSTONE;
    }

    @Override
    public IRegistration thenNotify(CancelTokenListener action) {
        return TOMBSTONE;
    }

    @Override
    public IRegistration thenNotifyAsync(Executor executor, CancelTokenListener action) {
        return TOMBSTONE;
    }

    @Override
    public IRegistration thenNotifyAsync(Executor executor, CancelTokenListener action, int options) {
        return TOMBSTONE;
    }

    @Override
    public IRegistration thenTransferTo(ICancelTokenSource child, int options) {
        return TOMBSTONE;
    }

    @Override
    public IRegistration thenTransferTo(ICancelTokenSource child) {
        return TOMBSTONE;
    }

    @Override
    public IRegistration thenTransferToAsync(Executor executor, ICancelTokenSource child) {
        return TOMBSTONE;
    }

    @Override
    public IRegistration thenTransferToAsync(Executor executor, ICancelTokenSource child, int options) {
        return TOMBSTONE;
    }
    // endregion
}

File: csharp/Wjybxx.Commons.Concurrent/src/Concurrent/UncancellableToken.cs
#region LICENSE

// Copyright 2023-2024 wjybxx(845740757@qq.com)
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#endregion

using System;

namespace Wjybxx.Commons.Concurrent;

/// <summary>
/// 不可取消的令牌
/// </summary>
sealed class UncancellableToken : ICancelToken
{
    public static readonly UncancellableToken Inst = new UncancellableToken();

    public ICancelToken AsReadonly => this;

    #region token

    public int CancelCode => 0;

    public bool IsCancelling() => false;

    public int Reason() => 0;

    public int Degree() => 0;

    public bool IsInterruptible() => false;

    public bool IsWithoutRemove => false;

    public void CheckCancel() {
    }

    #endregion

    #region 监听器

    public IRegistration ThenAccept(Action<ICancelToken> action, int options = 0) {
        return TOMBSTONE;
    }

    public IRegistration ThenAcceptAsync(IExecutor executor, Action<ICancelToken> action, int options = 0) {
        return TOMBSTONE;
    }

    public IRegistration ThenAccept(Action<ICancelToken, object?> action, object? ctx, int options = 0) {
        return TOMBSTONE;
    }

    public IRegistration ThenAcceptAsync(IExecutor executor, Action<ICancelToken, object?> action, object? ctx, int options = 0) {
        return TOMBSTONE;
    }

    public IRegistration ThenRun(Action action, int options = 0) {
        return TOMBSTONE;
    }

    public IRegistration ThenRunAsync(IExecutor executor, Action action, int options = 0) {
        return TOMBSTONE;
    }

    public IRegistration ThenRun(Action<object?> action, object? ctx, int options = 0) {
        return TOMBSTONE;
    }

    public IRegistration ThenRunAsync(IExecutor executor, Action<object?> action, object? ctx, int options = 0) {
        return TOMBSTONE;
    }

    public IRegistration ThenNotify(ICancelTokenListener action, int options = 0) {
        return TOMBSTONE;
    }

    public IRegistration ThenNotifyAsync(IExecutor executor, ICancelTokenListener action, int options = 0) {
        return TOMBSTONE;
    }

    public IRegistration ThenTransferTo(ICancelTokenSource child, int options = 0) {
        return TOMBSTONE;
    }

    public IRegistration ThenTransferToAsync(IExecutor executor, ICancelTokenSource child, int options = 0) {
        return TOMBSTONE;
    }

    #endregion

    private static readonly IRegistration TOMBSTONE = new MockRegistration();

    private class MockRegistration : IRegistration
    {
        public void Dispose() {
        }
    }
}